[core] Allow selecting packages to update via CRM or embark

fnussbaum opened the pull request:
Previously, selecting specific packages to update meant having to answer `yes-or-no-p` for each one. While a `list-packages` type of interface might be even better, I think this should be a clear improvement.

However, making selection work via embark requires some gymnastics (not sure if this is controversial).

fnussbaum committed:
[core] Split configuration-layer/update-packages into two functions

This allows programmatically updating specific packages (I am using this for
testing #16938, for example). It also prepares the possibility of selecting
packages to update through embark. Additionally, it might be slightly easier to
read.

fnussbaum committed:
configuration-layer/update-packages: Small refactors

Includes a tiny behavioural fix: We won't claim that "all packages are up to
date" when some could have been updated, but none were selected to update.

fnussbaum committed:
configuration-layer/update-packages: More small refactors

fnussbaum committed:
[core] Allow selecting packages to update via CRM or embark

Previously, selecting specific packages to update meant having to answer
`yes-or-no-p` for each one. While a `list-packages` type of interface might be
even better, I think this should be a clear improvement.

bcc32 commented on core/core-configuration-layer.el:
```diff
@@ -2102,64 +2099,88 @@ to update."
+             ;; order to avoid introducing another global variable (and because
+             ;; embark constrains the function to a single argument).
+             (require 'crm) ; for `crm-separator'
+             (fset 'configuration-layer//select-packages-to-update
```
Could this maybe use `cl-letf` with `symbol-function` as a generalized variable, so that the symbol doesn't have a leftover function definition once this is done?

fnussbaum replied:
Unfortunately, the function definition needs to extend beyond the scope of `configuration-layer/update-packages`, because embark-act-all calls the function from a post-command-hook.

(But I actually mildly liked being able to call `configuration-layer//select-packages-to-update` again to immediately update further packages that I missed before, without having to wait for the archives getting refreshed and all of the quelpa checkouts.)

bcc32 replied:
> Unfortunately, the function definition needs to extend beyond the scope of `configuration-layer/update-packages`, because embark-act-all calls the function from a post-command-hook.
Got it, thanks.

> being able to call `configuration-layer//select-packages-to-update` again to immediately update further packages that I missed before

IMO, this seems to argue in favor of just defining the command globally, and just using a global variable to stash the packages-to-update state.

Defining the command using `fset` makes `find-function` not work.

fnussbaum replied:
> IMO, this seems to argue in favor of just defining the command globally, and just using a global variable to stash the packages-to-update state.

Makes sense, I have pushed this, along with some minor changes.

bcc32 commented on core/core-configuration-layer.el:
```diff
@@ -2102,64 +2099,88 @@ to update."
+                   (lambda (selected-packages)
+                     (interactive (list (completing-read-multiple
+                                         (concat "Packages to update"
+                                                 (when (string-match-p crm-separator ",") " (comma separated)")
```
Is this necessary?  Spacemacs already has advice for updating the prompt for completing-read-multiple (see `crm-indicator`).  And besides, we're prompting for a plural.

fnussbaum replied:
Right, so perhaps this should rather be part of this advice (and the advice moved to spacemacs-defaults?). I'll consider creating a separate PR for this.

(It seems that `completing-read-multiple` is used only once within Spacemacs (and I don't think it is commonly used in packages, either), and it is not obvious how to use it the first time. I remember that this comma-separated indication helped me in the past.)

bcc32 commented on core/core-configuration-layer.el:
```diff
@@ -2102,64 +2099,88 @@ to update."
+                                         update-packages nil t)))
+                     (configuration-layer//update-packages
+                      (mapcar #'intern selected-packages))))
+             (let ((this-command 'configuration-layer//select-packages-to-update))
```
The `this-command` thing is mysterious, perhaps it deserves a comment.

fnussbaum replied:
Sure, added one now.

bcc32 reviewed (commented):
Thanks for keeping these commits separated, this was pretty easy to review.

One question, maybe subjective: I've been doing squash merges for most PRs but perhaps doing a rebase-then-non-ff-merge is warranted for more complicated changes?  I see that Spacemacs has mostly had a linear history but I already broke that once recently.  Opinions?  @smile13241324, do you have an opinion on squashing-vs-merging?

smile13241324 commented:
Not really @bcc32 our convention says we prefer rebasing over merging however we could change that. I personally think that sometimes not squashing all prs will make things more readable especially if it is not just micro changes.

bcc32 commented:
If merge commits are not desired we can always rebase the PR but without squashing, and then fast-forward into the `develop` branch.

fnussbaum commented:
Good idea, I would prefer not to squash sometimes, too.

fnussbaum committed:
configuration-layer/update-packages: Sort, remove frozen packages

Since we already print `update-packages` in alphabetical order, we might as well
keep it that way. Also, frozen packages will not be updated, even if selected,
so they should not be offered as an option.

bcc32 reviewed (approved):
Thanks, added a small comment but otherwise LGTM.

bcc32 commented:
(Merged without squashing)
